### PR TITLE
Add support for SSO bearer token authentication to the SDK

### PR DESCRIPTION
**Pull request by jdisanti**
This PR adds support for SSO bearer token authentication to the AWS SDK, specifically for Code Catalyst, which requires authentication via SSO with a Builder ID using a bearer token rather than SigV4.

This functionality was developed in a feature branch, and this PR merely merges that branch to main. The changes consist of the following previous PRs:
- https://github.com/smithy-lang/smithy-rs/pull/3381
- https://github.com/smithy-lang/smithy-rs/pull/3442
- https://github.com/smithy-lang/smithy-rs/pull/3443

All these changes have been reviewed in the previous PRs, but it would be good to review this again as a whole to verify it all looks good.

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: aws/sdk/integration-tests/codecatalyst/tests/sso_bearer_auth.json [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            "connection_id": 0,
+            "action": {
+                "Eof": {
+                    "ok": true,
+                    "direction": "Response"
+                }
+            }
+        }
+    ],
+    "docs": "todo docs",
```
Comment (ysaito1001) [inline aws/sdk/integration-tests/codecatalyst/tests/sso_bearer_auth.json:124]: It'd be good to edit `todo docs`.

**Review by ysaito1001 — APPROVED** [review]
Ship it 🚀

File: aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/TokenProvidersDecorator.kt [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        writable {
+            when (section) {
+                ServiceConfig.BuilderImpl -> {
+                    rustTemplate(
+                        """
+                        /// Sets the access token provider for this service
+                        ///
+                        /// Note: the [`Self::bearer_token`] and [`Self::bearer_token_resolver`] methods are
+                        /// equivalent to this method, but take the [`Token`] and [`ResolveIdentity`] types
+                        /// respectively.
```
Comment (ysaito1001) [inline aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/TokenProvidersDecorator.kt:86]: Thanks for adding these notes!

File: aws/rust-runtime/aws-config/src/profile/token.rs [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    /// let provider = ProfileFileTokenProvider::builder()
+    ///     .configure(&ProviderConfig::with_default_region().await)
+    ///     .build();
+    /// # }
+    /// ```
+    pub fn configure(mut self, provider_config: &ProviderConfig) -> Self {
+        self.provider_config = Some(provider_config.clone());
+        self
+    }
+
```
Comment (rcoh) [inline aws/rust-runtime/aws-config/src/profile/token.rs:147]: I missed this in first reading. We shouldn't expose new functions that expose `ProviderConfig` publicly. Prefer `SdkConfig` or have this method be private.

File: aws/rust-runtime/aws-config/src/profile/token.rs [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    /// make_test!(update: test_name)
+    /// ```
+    ///
+    /// **Run the test case against a real HTTPS connection:**
+    /// > Note: Be careful to remove sensitive information before committing. Always use a temporary
+    /// > AWS account when recording live traffic.
+    /// ```no_run
+    /// make_test!(live: test_name)
+    /// ```
+    macro_rules! make_test {
```
Comment (rcoh) [inline aws/rust-runtime/aws-config/src/profile/token.rs:200]: should we give this macro a different name?

File: aws/rust-runtime/aws-config/src/sso/token.rs [hunk 5]
```diff
@@ -466,15 +487,15 @@ mod tests {
...
-                "expected {err:?} to be `ExpiredToken`"
-            );
+            let err = DisplayErrorContext(
+                &self
+                    .token_provider
+                    .resolve_token(self.time_source.clone())
+                    .await
+                    .expect_err("expected failure"),
+            )
+            .to_string();
```
Comment (rcoh) [inline aws/rust-runtime/aws-config/src/sso/token.rs:497]: we should make an extension trait that does this

File: aws/rust-runtime/aws-config/src/profile/token.rs (hunk 3 above)
Comment (jdisanti) [reply aws/rust-runtime/aws-config/src/profile/token.rs]: Fixed in 89a2ea8895f32066e1ad803ec634683c307b0c84

File: aws/sdk/integration-tests/codecatalyst/tests/sso_bearer_auth.json (hunk 1 above)
Comment (jdisanti) [reply aws/sdk/integration-tests/codecatalyst/tests/sso_bearer_auth.json]: Fixed in 89a2ea8895f32066e1ad803ec634683c307b0c84

**Pull request merged**